In [0]:
import pandas as pd
import numpy as np
import warnings
import re

warnings.filterwarnings('ignore')

### Replace NA values with 0

In [0]:
stats_data = pd.read_csv("stats_Full.csv", na_values = "-").fillna("0")
stats_data

Unnamed: 0  Player_Id  ...  Attribute value
0           17259          0  ...      Squad  25.0
1           17259          0  ...      Squad   4.0
2           17259          0  ...      Squad   7.0
3           17259          0  ...      Squad   1.0
4           17259          0  ...      Squad   3.0
...           ...        ...  ...        ...   ...
62798      143798        168  ...      Squad   1.0
62799      143798        168  ...      Squad   3.0
62800      143798        168  ...      Squad   2.0
62801      143798        168  ...      Squad  25.0
62802      143798        168  ...          0     0

[62803 rows x 9 columns]

In [0]:
stats_data.dtypes

Unnamed: 0      int64
Player_Id       int64
tm_Id           int64
Name           object
Team           object
Season         object
Competition    object
Attribute      object
value          object
dtype: object

In [0]:
def clean_stats_value(row):
    value = row.value
    if(row.Attribute == "Minutes played" or row.Attribute == "Minutes per goal"):
      value = int(str(value).replace('.', ""))
      print(value)
    elif(row.Attribute == "Appearances" or row.Attribute == "Squad"):
      value = int(str(value).replace(".0", ""))
      print(value)
    elif(row.Attribute == "PPG"):
      value = float(value)/100
      print(value)
    return value

stats_data["value"] = stats_data['value'].str.replace('[^.\w\s]','')
# stats_data["value"] = stats_data['value'].str.replace('.0','')
# stats_data["value"] = pd.to_numeric(stats_data["value"], errors = "coerce")
noyer = stats_data[stats_data.tm_Id == 17259]
noyer_minutes = noyer[noyer.Attribute == "Minutes played"]
noyer_minutes
noyer

In [0]:
# stats_data["value"]
stats_data["value"] = stats_data['value'].str.replace('[^.\w\s]','')
stats_data["value"] = stats_data.apply(clean_stats_value, axis = 1)
# stats_data["value"] = stats_data['value'].str.replace('.0','')
# stats_data["value"] = pd.to_numeric(stats_data["value"], errors = "coerce")

In [0]:
stats_data

Unnamed: 0  Player_Id  ...  Attribute value
0             17259          0  ...      Squad  25.0
1             17259          0  ...      Squad   4.0
2             17259          0  ...      Squad   7.0
3             17259          0  ...      Squad   1.0
4             17259          0  ...      Squad   3.0
...             ...        ...  ...        ...   ...
1523706      148367       4719  ...    Assists   0.0
1523707      148367       4719  ...    Assists   0.0
1523708      148367       4719  ...  Own goals   0.0
1523709      148367       4719  ...  Own goals   0.0
1523710      148367       4719  ...          0   0.0

[1523711 rows x 9 columns]

In [0]:
noyer = stats_data[stats_data.tm_Id == 18723]
noyer_minutes = noyer[noyer.Attribute == "Minutes played"]
# noyer_minutes
noyer_minutes

Empty DataFrame
Columns: [Unnamed: 0, Player_Id, tm_Id, Name, Team, Season, Competition, Attribute, value]
Index: []

In [0]:
ppg = stats_data[stats_data.Attribute == "Goals"]
ppg[ppg.Name == "Cristiano Ronaldo"]

Unnamed: 0  Player_Id  ...  Attribute value
186564        8198        533  ...      Goals  21.0
186565        8198        533  ...      Goals   2.0
186566        8198        533  ...      Goals   2.0
186567        8198        533  ...      Goals   0.0
186568        8198        533  ...      Goals  21.0
...            ...        ...  ...        ...   ...
186631        8198        533  ...      Goals   0.0
186632        8198        533  ...      Goals   3.0
186633        8198        533  ...      Goals   2.0
186634        8198        533  ...      Goals   0.0
186635        8198        533  ...      Goals   0.0

[72 rows x 9 columns]

In [0]:
stats_data = stats_data.drop("Unnamed: 0", axis = 1)
stats_data

Player_Id   tm_Id  ...  Attribute value
0                0   17259  ...      Squad  25.0
1                0   17259  ...      Squad   4.0
2                0   17259  ...      Squad   7.0
3                0   17259  ...      Squad   1.0
4                0   17259  ...      Squad   3.0
...            ...     ...  ...        ...   ...
1523706       4719  148367  ...    Assists   0.0
1523707       4719  148367  ...    Assists   0.0
1523708       4719  148367  ...  Own goals   0.0
1523709       4719  148367  ...  Own goals   0.0
1523710       4719  148367  ...          0   0.0

[1523711 rows x 8 columns]

In [0]:
# Function checks whether given column's rows contain any special characters,
# and returns those rows from the dataframe.

def check_string(column):
    col = column.str.contains('[@_!#$%^&*()<>?/\|}{~:]')
    idx = list(np.where(col == True)[0])
    rows = column.iloc[idx]
    
    if len(rows) == 0:
        return "Column does not contain special characters"
    else:
        return rows

In [0]:
check_string(stats_data.Name)

'Column does not contain special characters'

In [0]:
stats_data.sort_values("Player_Id", axis = 0, ascending = True, inplace = True)
stats_data

Player_Id   tm_Id  ...            Attribute  value
0                0   17259  ...                Squad  25.00
483              0   17259  ...  Second yellow cards   0.00
484              0   17259  ...  Second yellow cards   0.00
485              0   17259  ...  Second yellow cards   0.00
486              0   17259  ...  Second yellow cards   0.00
...            ...     ...  ...                  ...    ...
1523616       4719  148367  ...          Appearances   0.00
1523617       4719  148367  ...          Appearances  16.00
1523618       4719  148367  ...          Appearances   7.00
1523638       4719  148367  ...                  PPG   1.75
1523710       4719  148367  ...                    0   0.00

[1523711 rows x 8 columns]

In [0]:
stats_data = stats_data.drop("Player_Id", axis = 1)

tm_Id                Name  ...            Attribute  value
0         17259        Manuel Neuer  ...                Squad  25.00
483       17259        Manuel Neuer  ...  Second yellow cards   0.00
484       17259        Manuel Neuer  ...  Second yellow cards   0.00
485       17259        Manuel Neuer  ...  Second yellow cards   0.00
486       17259        Manuel Neuer  ...  Second yellow cards   0.00
...         ...                 ...  ...                  ...    ...
1523616  148367  Christian Nörgaard  ...          Appearances   0.00
1523617  148367  Christian Nörgaard  ...          Appearances  16.00
1523618  148367  Christian Nörgaard  ...          Appearances   7.00
1523638  148367  Christian Nörgaard  ...                  PPG   1.75
1523710  148367  Christian Nörgaard  ...                    0   0.00

[1523711 rows x 7 columns]

In [0]:
stats_data = stats_data[stats_data.Season != '0']
stats_data.Season.unique()

array(['19/20', '05/06', '04/05', '03/04', '18/19', '17/18', '16/17',
       '06/07', '11/12', '10/11', '09/10', '08/09', '07/08', '15/16',
       '14/15', '13/14', '12/13', '2016', '2018', '2017', '2010', '2009',
       '2015', '2014', '2011', '2013', '2012', '2008', '2019', '2006',
       '2007', '2002', '2003', '2004', '2005', '02/03', '99/00', '01/02',
       '00/01', '95/96', '97/98', '96/97', '98/99', '2019.0', '2018.0',
       '2017.0', '2016.0', '92/93', '91/92', '93/94', '94/95', '90/91',
       '88/89', '89/90', '2020'], dtype=object)

In [0]:
stats_data = stats_data.reset_index(drop = True)
stats_data

tm_Id                Name  ...            Attribute  value
0         17259        Manuel Neuer  ...                Squad  25.00
1         17259        Manuel Neuer  ...  Second yellow cards   0.00
2         17259        Manuel Neuer  ...  Second yellow cards   0.00
3         17259        Manuel Neuer  ...  Second yellow cards   0.00
4         17259        Manuel Neuer  ...  Second yellow cards   0.00
...         ...                 ...  ...                  ...    ...
1523705  148367  Christian Nörgaard  ...          Appearances  13.00
1523706  148367  Christian Nörgaard  ...          Appearances   0.00
1523707  148367  Christian Nörgaard  ...          Appearances  16.00
1523708  148367  Christian Nörgaard  ...          Appearances   7.00
1523709  148367  Christian Nörgaard  ...                  PPG   1.75

[1523710 rows x 7 columns]

### Drop no_info names

In [0]:
no_names_idx = stats_data.loc[stats_data.Name == "no_info"].index
stats_data = stats_data.drop(no_names_idx)

check_string(stats_data.Name)

'Column does not contain special characters'

### Drop Name column

In [0]:
stats_data = stats_data.drop("Name", axis = 1)
stats_data = stats_data.reset_index(drop = True)


KeyError: ignored

In [0]:
stats_data.loc[stats_data.Season == "19/20"]

tm_Id           Team Season       Competition  Attribute  value
0         17259  Bayern Munich  19/20        Bundesliga      Squad  25.00
6         17259  Bayern Munich  19/20        Bundesliga  Red cards   0.00
7         17259  Bayern Munich  19/20         DFB-Pokal  Red cards   0.00
8         17259  Bayern Munich  19/20  Champions League  Red cards   0.00
10        17259  Bayern Munich  19/20      DFL-Supercup  Red cards   0.00
...         ...            ...    ...               ...        ...    ...
1523658  148367   Brentford FC  19/20            FA Cup  Own goals   0.00
1523659  148367   Brentford FC  19/20      Championship      Squad  33.00
1523660  148367   Brentford FC  19/20            FA Cup      Squad   1.00
1523677  148367   Brentford FC  19/20      Championship        PPG   1.64
1523678  148367   Brentford FC  19/20            FA Cup        PPG   0.00

[186712 rows x 6 columns]

### Questions
* How to categorize "Season" feature?

In [0]:
# stats_data.to_csv(r"full_data/Cleaned_stats_Full.csv", index = False)
print(stats_data.dtypes)
stats_data.to_pickle("Cleaned_stats_Full.pkl")

tm_Id            int64
Team            object
Season          object
Competition     object
Attribute       object
value          float64
dtype: object
